### ISSUE TITLE
AVFoundation.AVSpeechSynthesizer Bug when using prefersAssistiveTechnologySettings Property

### ISSUE THREAD

**Issue by tbreitenfeldt** [ronaldoussoren/pyobjc#439]
When creating an AVFoundation.AVSpeechSynthesizer object and using the prefersAssistiveTechnologySettings Property, user Voiceover settings are not applied. The code runs, however the Voiceover uses the default speech settings to speak the provided string even though prefersAssistiveTechnologySettings is set to True. The expected result is for Voiceover to speak the utterance using the users existing settings for Voiceover I.E. rate, pitch, voice, etc... when prefersAssistiveTechnologySettings is true.

Code sample is below:

```
from AVFoundation import AVSpeechSynthesizer
from AVFoundation import AVSpeechUtterance

voiceover = AVSpeechSynthesizer.alloc().init()
utterance = AVSpeechUtterance.speechUtteranceWithString_("hello world")
utterance.setPrefersAssistiveTechnologySettings_(True)
voiceover.speakUtterance_(utterance)
input()
```

Here is the Apple documentation for the AVSpeechSynthesizer
[https://developer.apple.com/documentation/avfaudio/avspeechsynthesizer?language=objc](https://developer.apple.com/documentation/avfaudio/avspeechsynthesizer?language=objc)
and the link to the pypi for the Python bindings
(https://pypi.org/project/pyobjc-framework-AVFoundation/)[https://pypi.org/project/pyobjc-framework-AVFoundation/]

**Comment by ronaldoussoren** [comment]
I expect that AVSpeechSynthesizer requires a runloop to do its work. That said, just replacing "input()" by "NSRunLoop.currentRunLoop().run()" doesn't fix the issue.

I've never used this API myself and will have to do some more research to be sure.

**Comment by tbreitenfeldt** [comment]
Hi, thanks for the quicvk response, I don't think so since speech synthesizers are supposed to be asynchronous. The input() at the end is to insure that Voiceover has the chance to speak before the python application terminates.
I gave it a try though, and got no change, other than being unable to kill the python app without killing the terminal. Here is the updated code that I used

```
from AVFoundation import AVSpeechSynthesizer
from AVFoundation import AVSpeechUtterance
from Foundation import NSRunLoop

voiceover = AVSpeechSynthesizer.alloc().init()
utterance = AVSpeechUtterance.speechUtteranceWithString_("hello world")
utterance.setPrefersAssistiveTechnologySettings_(True)
voiceover.speakUtterance_(utterance)
NSRunLoop.currentRunLoop().run()
```

Honestly, I am not very familiar with ObjectiveC or pyobjc, although I believe this is what I need for a solution for a project I am working on. I was wondering if booleans in pyobjc are handled differnetly, and if the line 
`utterance.setPrefersAssistiveTechnologySettings_(True)`
should be something different. Is it necessary to define the boolean using a pyobjc object? I have been doing research and nothing is helping much.

**Comment by ronaldoussoren** [comment]
I've added a delegate to the script:

```python
class Delegate(NSObject):
    def speechSynthesizer_didStartSpeechUtterance_(self, a, b):
        print("didStart", a, b)

    def speechSynthesizer_didCancelSpeechUtterance_(self, a, b):
        print("didCancel", a, b)

    def speechSynthesizer_willSpeakRangeOfSpeechString_utterance_(self, synth, value, utt):
        print("Will speak", synth, value, utt)

    def speechSynthesizer_didFinishSpeechUtterance_(self, synth, utterance):
        print("done", synth, utterance)
        sys.exit(42)

...

delegate = Delegate.alloc().init()
VoiceOver.setDelegate_(delegate)
```

That shows that the utterance is cancelled immediately, the other delegate methods do not get called.

I don't know why this is, maybe the code needs

**Comment by ronaldoussoren** [comment]
That wasn't it, you need to set the voice for the utterance after creating it:

```python
utterance.setVoice_(AVSpeechSynthesisVoice.voiceWithLanguage_("en-GB"))
```

This ensures that the text is spoken, and that doesn't require having a runloop.

The full script:

```python
from AVFoundation import AVSpeechSynthesizer
from AVFoundation import AVSpeechUtterance
from AVFoundation import AVSpeechSynthesisVoice
import time

voiceover = AVSpeechSynthesizer.alloc().init()
utterance = AVSpeechUtterance.speechUtteranceWithString_("hello world")
utterance.setPrefersAssistiveTechnologySettings_(True)
utterance.setVoice_(AVSpeechSynthesisVoice.voiceWithLanguage_("en-GB"))
voiceover.speakUtterance_(utterance)

# Sleep for a bit to wait for the text to be spoken
# This isn't the best way to do this, it would be
# better to sleep in a loop and check if the synthesizer
# is done yet every time through the loop.
time.sleep3) 
```

**Comment by tbreitenfeldt** [comment]
I see, although this is not quite what I am looking for. In the Apple documentation for
(AVSpeechUtterance)[https://developer.apple.com/documentation/avfaudio/avspeechutterance?language=objc]
It states here that this setting is supposed to use the users existing Voiceover settings in the utterance. This is to allow for a more seamless experience for Voiceover users. Users can go from using Safari, to using what ever application has implemented this, and the user should not notice any difference in speech output. I have my Voiceover settings modified from the default, using a different voice, and rate. I would expect that when AVSpeechSynthesizer speaks, that the voicwe and rate I have set in Voiceover is used.

From the NSSpeechUtterance docs:
> prefersAssistiveTechnologySettings
> A Boolean that specifies whether assistive technology settings take precedence over the property values of this utterance.

From everything I keep reading this is how this is supposed to behave, however I am having challenges creating an application in swift or objectiveC to test this properly.

**Comment by tbreitenfeldt** [comment]
I would like to request that this issue is reopened since the original problem was not addressed. IN the closing comment, setting the voice does not address the issue of getting speech output using existing Voiceover user custom settings.

**Comment by ronaldoussoren** [comment]
I'm pretty sure that the issue you run into is not related to PyObjC, but in the way the API is used.

Please share a complete script and additional information on how to reproduce the problem.  The script you shared initially does not emit speech at all, at least not on my machine (running macOS 12) and I don't know how to check if the speech synthesizer uses changed VoiceOver settings (or even how to change those settings).

**Comment by tbreitenfeldt** [comment]
Hi, thank you for reopening the issue. That is strange that you got no speech output with the first script I posted. My appologies for not describing Voiceover well enough, I am just used to using it.
Voiceover can be ran by opening it from your utilities folder. Voiceover is a built in application for blind and visually impaired users. The shortcut for enabling and disabling Voiceover is command+f5. The first time you launch Voiceover I think it will pull up the settings dialog to change default settings, if not, you can dig around I think in utilities to find voiceover settings, or just use the shortcut control+option+f8. In preferences you can change the voice that is speaking, the rate, pitch, and a number of other settings. Most blind people do not leave the default settings for Voiceover, often increasing the speed in which Voiceover talks to improve productivity. From my understanding of how the AVSpeechUtterance.prefersAssistiveTechnologySettings property is supposed to work is that rather than speaking with the default voice, or even with any configurations made to the utterance object, the custom user settings from Voiceover are supposed to take precedent.

I was finally able to get a Swift script put together to test AVSpeechSynthesizer though, and unfortunately did not have much more luck. I still got speech output with the default settings rather than my custom set settings. This might seem like a nitpicky thing to be working for, but screen reader users get very used to working with the settings they set, and if someone uses this library in say their desktop application, it will be very jarring for the user if the usual  voice they were just using to browse the internet is different when they use this desktop program. 

Anyways, below is the Swift code I came up with. Given that this is not working in swift as expected, I must assume that this is not a pyobjc issue, and is perhaps related to me not understanding how this property is actually functioning, or perhaps the underlying API is broken. There is just not enough information on these APIs for speech output for Voiceover, so it is difficult to say.

```
import Foundation
import AVFoundation

if #available(macOS 10.14, *) {
    let voiceover = AVSpeechSynthesizer()
    let utterance = AVSpeechUtterance(string: "hello world!")
    if #available(macOS 11.0, *) {
        print("sets prefersAssistiveTechnologySettings")
        utterance.prefersAssistiveTechnologySettings = true
    }
    voiceover.speak(utterance)
    let _ = readLine()
}
```

Timothy Breitenfeldt

**Comment by ronaldoussoren** [comment]
Its probably best to ask about this on stack overflow or the apple developer forums.  With some luck an Apple engineer will result on the latter.   I'd use your Swift snippet in the question, and not Python, as that's more likely to result in an answer.
